### PR TITLE
[`KTOTrainer`] add BCO (reward shift and underlying distribution matching)

### PR DESCRIPTION
**Pull request by seanexp**
add Binary Classifier Optimization (BCO) loss function from https://arxiv.org/abs/2404.04656

Implemented BCE loss, reward shift and underlying distribution matching.

Also added example script at `examples/scripts/bco.py`

### CODE REVIEW

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@seanexp  can you kindly run:

```
pre-commit run --all-file
```
in the root folder of TRL to clean up the formatting?

**Comment by seanexp** [comment]
@kashif 

My bad. Applied `pre-commit` at [a9a7a9c](https://github.com/huggingface/trl/pull/1599/commits/a9a7a9c0093d5a6da5b5ab3ae16f8fc999ffb6ed)

**Comment by kashif** [comment]
thanks @seanexp one questions, would it make sense in the example script to have an option to try with different embedding models? or are things hard-coded for nomic-embed?

**Comment by kashif** [comment]
@seanexp  you might need to add scikit-learn dep here: https://github.com/huggingface/trl/blob/main/setup.py#L72

**Comment by kashif** [comment]
oh sorry @seanexp  i thought the scikit-learn is in the tests but its in the main trainer... so we need to add that as a regular dependency! 🙇🏽

**Comment by seanexp** [comment]
@kashif 

I found that adding `scikit-learn` is redundant as it is a required package of `transformers`.

Should I still add `scikit-learn` as a regular dependency?

**Comment by kashif** [comment]
@seanexp yeah lets remove the dep from the test then... you can see here that even though transformers is installed, the tests fail: https://github.com/huggingface/trl/actions/runs/8891165233/job/24412685192#step:5:2843

**Comment by seanexp** [comment]
Seems like the dep in the test is necessary. Without the dep the test fails. Shall we leave as it is? @kashif

**Comment by kashif** [comment]
@seanexp ok lets leave the import in the tests and add a helper here: https://github.com/huggingface/trl/blob/main/trl/import_utils.py for scikit-learn and then in the trainer we can check if its available when BCO is the loss type and if not, we can ask that they install it via `pip instal ...`

**Comment by kashif** [comment]
@seanexp  something like this in the KTOConfig: https://github.com/huggingface/trl/blob/main/trl/trainer/ddpo_config.py#L116-L120

**Comment by seanexp** [comment]
@kashif 

done at [ddbbbdf](https://github.com/huggingface/trl/pull/1599/commits/ddbbbdf7624e715b586cd805ad9bb1b65db22f7c) :)

**Comment by seanexp** [comment]
> thanks @seanexp one questions, would it make sense in the example script to have an option to try with different embedding models? or are things hard-coded for nomic-embed?

Users can try different embedding models. Please note that users have to modify `embedding_func` (and `embed_prompt`) accordingly as embedding models asks different post processing methods (e.g. mean/max pooling, using only subset of features) @kashif

**Review by kashif — APPROVED** [review]
LGTM!

**Comment by seanexp** [comment]
@younesbelkada 

hmm... I'll take a look few hours later!

**Comment by seanexp** [comment]
Not calling `super().__post_init__()` was the problem. Fixed at [5c58065](https://github.com/huggingface/trl/pull/1599/commits/5c580655a2bebc3641fef688d5acb105ef083fc7)

The tests now run properly on my machine :)

**Comment by kashif** [comment]
good catch @seanexp 🥇

**Review by younesbelkada — APPROVED** [review]
Woah great catch ! Thanks again for adding this nice feature !
